doc/decisions/constructor_functions.md

```diff
@@ -0,0 +1,85 @@
+# Constructor Functions
+
+## Problem
+
+The constructor functions that `libelektra-core` defines for creating Keys and KeySets must be possible to implement in other languages without complicated workarounds.
+In particular, another implementation should not need to resort to adding a small amount of C to "bridge the gap".
+
+## Constraints
+
+- Constructor functions should use as little resources as feasible.
+  In particular, the number of `malloc`s and `memcpy` should be kept in check.
+  Temporary allocations during key creation should be avoided.
+- Using the escaped name in the constructor function should be avoided, since the unescaping process can be expensive.
+
+## Assumptions
+
+- The constructor functions from `libelektra-core` will not or only very rarely be used directly by users.
+- For all languages supported by Elektra (including C), there is secondary library on top of `libelektra-core` that provides more idiomatic APIs.
+- Creating a key with a name, no value and no metadata and then calling `keySetValue`/`keySetMeta` does not create significant overhead compared to creating the key with value and metadata directly.
+
+  > **Note:** The current `keyNew` just calls `keySetValue`/`keySetMeta` internally.
+
+## Considered Alternatives
+
+- Create a constructor function based on heavy use of arrays and `NULL` terminators
+- A single function that takes `char * name`, `void * value` and `KeySet * meta` for Keys or a `Key **` for KeySets.
+  Sizes would need to be passed as separate arguments.
+
+## Decision
+
+`libelektra-core` only provides the absolute minimal API to create a `Key *` or `KeySet *`:
+
+```c
+/**
+ * Allocates a new key and returns it.
+ * The key will be in the namespace @p ns.
+ *
+ * If `name == NULL`, the returned key is the root key of @p ns and @p nameSize is ignored.
+ * If `nameSize == 0`, the returned key is the root key of @p ns and @p name is ignored.
+ * Otherwise, the name of the key within the namespace will be @p name of length @p nameSize (including null terminator).
+ * Therefore, the returned key `k` fullfills:
+ * `keyNameSize (k) == nameSize + 2`
+ * (one byte for the namespace, and one for the separator between namespace and the rest of the name)
+ *
+ *
+ * The key will have no value and no metadata.
+ */
+Key * elektraKeyNew (elektraNamespace ns, const char * name, size_t nameSize);
+
+/**
+ * Allocates a new keyset with space reserved for at least @p keyCount keys.
+ * If `keyCount == 0`, the keyset will be empty (with no space for keys reserved) and @p keys is ignored.
+ * If `keys == NULL`, the keyset will be empty with space reserved for at least @p keyCount keys.
+ * Otherwise the keyset will contain all `keys[i]` for any `0 <= i < keyCount` where `keys[i] != NULL`.
+ * In other words, `NULL`s within @p keys will be ignored and if @p keys contains more than @p keyCount keys
+ * the additional ones will also be ignored.
+ */
+Key * elektraKeySetNew (Key ** keys, size_t keyCount);
+```
```

Comment:
I decided to use a slightly more useful API than before. I still think that this API (especially `elektraKeyNew`) should almost never be used directly. However, the previous API was essentially just an allocator and had the problem that you always create a key with the wrong name first.

The fact that the size must be passed separately is a bit annoying, but since it should be hidden by another API in practice it doesn't matter much. The API should also still be simple enough that it could be implemented in any language.

Reply:
Please write this in the decision (considered alternatives). It is unclear to me which text refers to Key and which to KeySet. In case of the Key** the size is actually not needed and it could be null-terminated.

> The API should also still be simple enough that it could be implemented in any language.

As said above, it was also possible to call keyNew from any language.